quinn/src/connection.rs

```diff
@@ -111,17 +107,19 @@ impl Future for ConnectionDriver {
             keep_going |= conn.handle_timer_updates();
             conn.forward_endpoint_events();
             conn.forward_app_events();
-            if !keep_going || conn.closed {
+            if !keep_going || conn.inner.is_drained() {
                 break;
             }
         }
 
         if !conn.inner.is_drained() {
-            Ok(Async::NotReady)
-        } else if let Some(e) = &conn.pending.error {
-            Err(e.clone())
-        } else {
-            Ok(Async::Ready(()))
+            conn.driver = Some(task::current());
+            return Ok(Async::NotReady);
+        }
+        match conn.pending.error {
```

Comment:
Shouldn't this go before the `if !conn.inner.is_drained()`, similar to your changes in `IncomingStreams::poll()`?

Reply:
It specifically must not, because `error` is usually set before the connection becomes drained. Everything but the driver should complete immediately at that point, but the driver *must* remain until draining is complete so that it can gracefully absorb incoming packets.

Future work could be to allow move the entire draining state into `Endpoint`, which would allow drivers to be dropped instantly as well, but that's not the current shape of things.

Reply:
Maybe add a comment then to that make that more obvious? Also, why would we want to move the draining state into the `Endpoint`?